go-controller/hybrid-overlay/pkg/controller/node_linux.go

```diff
@@ -0,0 +1,475 @@
+			if len(cookie) != 8 {
+				// Ignore non-node-specific rules (eg cookie=0x0)
+				continue
+			}
+			if _, ok := kubeNodes[cookie]; !ok {
+				nodesToRemove[cookie] = true
+			}
+		}
+	}
+
+	for cookie := range nodesToRemove {
+		if err := deleteNodeFlowsByCookie(cookie); err != nil {
+			klog.Errorf("Failed clean stale hybrid overlay node flow %q: %v", cookie, err)
+		}
+	}
+}
+
+func getLocalNodeSubnet(nodeName string) (*net.IPNet, error) {
+	var cidr string
+	var err error
+
+	// First wait for the node logical switch to be created by the Master, timeout is 300s.
+	if err := wait.PollImmediate(500*time.Millisecond, 300*time.Second, func() (bool, error) {
+		if cidr, _, err = util.RunOVNNbctl("get", "logical_switch", nodeName, "other-config:subnet"); err != nil {
+			return false, nil
+		}
+		return true, nil
+	}); err != nil {
+		return nil, fmt.Errorf("timed out waiting for node %q logical switch: %v", nodeName, err)
+	}
+
+	_, subnet, err := net.ParseCIDR(cidr)
+	if err != nil {
+		return nil, fmt.Errorf("Invalid hostsubnet found for node %s - %v", nodeName, err)
+	}
+
+	klog.Infof("found node %s subnet %s", nodeName, subnet.String())
+	return subnet, nil
+}
+
+func getIPAsHexString(ip net.IP) string {
+	if ip.To4() != nil {
+		ip = ip.To4()
+	}
+	asHex := ""
+	for i := 0; i < len(ip); i++ {
+		asHex += fmt.Sprintf("%02x", ip[i])
+	}
+	return asHex
+}
+
+func (n *NodeController) ensureHybridOverlayBridge() error {
+	var err error
+	if n.subnet, err = getLocalNodeSubnet(n.nodeName); err != nil {
+		return err
+	}
+
+	portName := houtil.GetHybridOverlayPortName(n.nodeName)
+
+	// If the master hasn't yet created our hybrid overlay port, try again later
+	portMAC, portIP, _ := util.GetPortAddresses(portName)
+	if portMAC == nil || portIP == nil {
+		return nil
+	}
+
+	// if our bridge already exists, nothing to do
+	if _, _, err := util.RunOVSVsctl("br-exists", extBridgeName); err == nil {
+		return nil
+	}
+
+	// Otherwise set things up
+	_, stderr, err := util.RunOVSVsctl("--may-exist", "add-br", extBridgeName,
+		"--", "set", "Bridge", extBridgeName, "fail_mode=secure")
+	if err != nil {
+		return fmt.Errorf("Failed to create localnet bridge %s"+
+			", stderr:%s: %v", extBridgeName, stderr, err)
+	}
+
+	// A OVS bridge's mac address can change when ports are added to it.
+	// We cannot let that happen, so make the bridge mac address permanent.
+	macAddress, err := util.GetOVSPortMACAddress(extBridgeName)
+	if err != nil {
+		return err
+	}
+	stdout, stderr, err := util.RunOVSVsctl("set", "bridge", extBridgeName, "other-config:hwaddr="+macAddress)
+	if err != nil {
+		return fmt.Errorf("Failed to set bridge, stdout: %q, stderr: %q, "+
+			"error: %v", stdout, stderr, err)
+	}
+
+	if _, _, err = util.RunIP("link", "set", extBridgeName, "up"); err != nil {
+		return fmt.Errorf("failed to up %s: %v", extBridgeName, err)
+	}
+
+	const (
+		rampInt string = "int"
+		rampExt string = "ext"
+		portNum string = "11"
```

Comment:
@danwinship IIRC the port numbers are per-OVS-bridge though, and since this is `br-ext` nobody is going to use it for something else unless they are using the hybrid overlay. I've changed the variable to `rampExtPortNum`